### ISSUE TITLE
Validating nomnoml diagrams?

### ISSUE THREAD

**Issue by javierluraschi** [javierluraschi/nomnoml#19]
Currently there is no way to validate if a `nomnoml` diagram is valid, say `[hello]` (valid), vs `[hello` (invalid).

However, we can use the [mlverse/duktape](https://github.com/mlverse/duktape) package as a workaround as follows:

```
remotes::install_github("mlverse/duktape")
```

Then define the following function:

```r
nomnoml_validate <- function(diagram = "[test]") {
    duktape::eval(
        paste(c(
            readLines(system.file("htmlwidgets/lib/dagre/dagre.min.js", package = "nomnoml")),
            readLines(system.file("htmlwidgets/lib/lodash/lodash.js", package = "nomnoml")),
            readLines(system.file("htmlwidgets/lib/nomnoml/nomnoml.js", package = "nomnoml")),
            paste0("nomnoml.parse('", diagram, "')")
        ), collapse = "\n")
    )
}
```

And validate `nomnoml` syntax as follows:

```r
nomnoml_validate("[hello]")
```

vs

```r
nomnoml_validate("[hello")
```
```
 Error in duktape::rcpp_eval(paste(c(readLines(system.file("htmlwidgets/lib/dagre/dagre.min.js",  : 
  Error: Parse error on line 1:
[hello
------^
Expecting 'SEP', '|', ']', got 'EOF' 
```

**Comment by andrie** [comment]
This is great, thank you!

I tried your code, and it works great for the example provided, but I can't get it to work for multi-line diagrams, e.g:

```r
nomnoml_validate("[hello]
                [world]")
```

results in:

```
Error in rcpp_eval(code) : SyntaxError: unterminated string (line 19004) 
```

However, I was able to get the same idea to work using `V8` instead of `duktape`:

```r
nomnoml_validate <- function(diagram = "[test]") {
  ct <- v8()
  ct$source(system.file("htmlwidgets/lib/dagre/dagre.min.js", package = "nomnoml"))
  ct$source(system.file("htmlwidgets/lib/lodash/lodash.js",   package = "nomnoml"))
  ct$source(system.file("htmlwidgets/lib/nomnoml/nomnoml.js", package = "nomnoml"))
  ct$assign("diagram", diagram)
  ct$eval("nomnoml.parse(diagram)"),
}
```

An invalid diagram:

```
> nomnoml_validate("[hello")
<std::runtime_error in context_eval(join(src), private$context, serialize): Error: Parse error on line 1:
[hello
------^
Expecting 'SEP', '|', ']', got 'EOF'>
```

A valid multi-line diagram:

```
> nomnoml_validate("[hello]
+                 [world]")
[1] "[object Object]"
```

Is there any disadvantage to use `V8` compared to `duktape`?

**Comment by andrie** [comment]
I implemented this in #21
